NeoML/src/Dnn/ReferenceDnnFactory.cpp

```diff
@@ -0,0 +1,145 @@
+﻿/* Copyright © 2024 ABBYY
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#include <common.h>
+#pragma hdrstop
+
+#include <NeoML/Dnn/Dnn.h>
+#ifdef NEOML_COMPACT // No optimizations in mobile assembly
+namespace NeoML {
+static void OptimizeDnn( CDnn& ) {}
+}
+#else  // !NEOML_COMPACT
+#include <NeoML/Dnn/DnnOptimization.h>
+#endif // !NEOML_COMPACT
+
+namespace NeoML {
+
+// Internal technical class
+class CReferenceDnnInfo final {
+public:
+	CReferenceDnnInfo( CRandom rand, CReferenceDnnFactory& ref ) : random( std::move( rand ) ), owner( ref )
+	{ owner.counter += 1; }
+
+	~CReferenceDnnInfo()
+	{ owner.counter -= 1; }
+
+	CRandom& Random() { return random; }
+
+private:
+	// Stores the dnn's own external random class inside this dnn class
+	CRandom random;
+	// The factory accounts the number of owned reference dnns
+	CReferenceDnnFactory& owner;
+};
+
+void CReferenceDnnInfoDeleter::operator()( CReferenceDnnInfo* info ) { if( info != nullptr ) { delete info; } }
+
+//---------------------------------------------------------------------------------------------------------
+
+CReferenceDnnFactory::CReferenceDnnFactory( IMathEngine& mathEngine, CArchive& archive, int seed, bool optimizeDnn ) :
+	CReferenceDnnFactory( MakeCPtrOwner<CReferenceDnnInfo>( CRandom( seed ), *this ), mathEngine )
```

Comment:
3 duplications of `MakeCPtrOwner` call -> just relocate it to technical ctor.

Reply:
By the way, such passings of `CPtrOwner` must use rvalue references.

Reply:
I will fix this when transfer to `CPtr` for factory